cataclysmbnteam/Cataclysm-BN
# feat(port): Add book binder with many QoL

#### Summary
SUMMARY: Features "Add book binder with many QoL"

#### Purpose of change
Being able to copy and keep the most interesting recipes on you, even if they were above your level when you copied them.
Also, it was one of the (many) PR to port before
https://github.com/CleverRaven/Cataclysm-DDA/pull/46304
and
https://github.com/CleverRaven/Cataclysm-DDA/pull/49408 (the final boss)

#### Describe the solution
Port https://github.com/CleverRaven/Cataclysm-DDA/pull/44399
and https://github.com/CleverRaven/Cataclysm-DDA/pull/50367

Changes from the original PR:
- when you copy a new recipe, the old recipes that you already know are removed with a message. This means you'll be able to keep the same book binder for the entire game.
![image](https://user-images.githubusercontent.com/71428793/200210313-b61f7ff5-0255-44b3-8275-b8c9cfdf6f03.png)

- the PR added some randomness to the spawn of paper, I didn't port it
- you could only copy recipes that match your skill level, I removed the requirement
  - but instead, I added a check on the skill level when you open your craft menu
  - which means, you can copy any recipe from any book, but you will only be able to craft it when you met the requirement
  - I also removed the learnt recipes, and the recipes already in the book binder from the selection
  - hopefully, those changes will make it more user friendly
- the pen can be used to write too (it has some charges now)
- the copying time has been greatly reduced (it was quite long before, probably aligned with DDA pace)
- lots of change to the book binder item
  - it's not a GENERIC item, but a TOOL
  - it uses paper as charges
  - by default (when you craft it, and when it spawns), it has respectively 0 or 1 charges (the first page is used)
  - then, copying a recipe increases the amount of charges in it
  - when it's full of charges, you can't add new recipes in this one
  - I could make it infinite, but since it's a tool and not a GENERIC item that contains paper, only the weight increases, not the volume. So I set the maximum at 500 pages (roughly 100 recipes). Although with the first point of this section, it doesn't matter much now.
  - lots of extra check have been added, or moved before the recipe to copy selection, to warn the player earlier that the action can't be done (cf tests)


#### Describe alternatives you've considered
Not porting it, and port the e-book one directly (but way too many changes)
Porting the e-book one in this PR too (too big)


#### Testing
- remove the flags UNLOAD, and RELOAD, spawn new character with 0 skills
- test the following conditions: book is full, no paper on player, no writing tool, in the dark, you get special messages before selecting a recipe to copy
- test the following conditions: 2 papers on the player, or "1 paper left" in the book: only the recipes with a number of page below or equal to those requirements can be selected (wasn't the case)
- spawn the "A History of Firefighting" book
- copy the fire axe and the crash axe recipe
- the book has now 5 + 4 pages in it
- open the copy recipe menu, they are not here anymore
- drop the "A History of Firefighting" book, and go away from it
- select view recipe, they are here
- increase your fabrication skill to 5, try to craft them, they don't appear
- increase your fabrication skill to 6, the crash axe appears
- increase your fabrication skill to 9, the fire axe appears
- drop the book binder, go away from it
- they don't appear in the craft menu (not memorised yet)
- read the "A History of Firefighting" book once
- try to copy a recipe, then view the recipes. They disappeared because you learnt them by reading the book
